[Fix] Auto-detect XGrammar compiler threads based on CPU cores.

Ubospica opened the pull request:
This PR detects the number of threads in XGrammar's GrammarCompiler with the number of CPU cores to achieve best compilation performance.

Thanks to discussion with @russellb

Ubospica committed:
finish

Signed-off-by: Ubospica <ubospica@gmail.com>

update

Signed-off-by: Ubospica <ubospica@gmail.com>

update comments

fix pre-commit errors

Signed-off-by: Ubospica <ubospica@gmail.com>

russellb commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -87,11 +88,22 @@ def __init__(self, vllm_config: VllmConfig):
+    def _get_grammar_compiler_threads(self) -> int:
+        """Get the number of threads to use for the grammar compiler. The number of physical
+        threads is used to determine the thread count, and it is capped at 32 because having too
+        many threads does not yield additional benefits.
```
Do you think there's something special about 32? or just that it happened to be the number of physical cores on a single CPU in the test environment?

Ubospica replied:
The maximum number of effective threads is the number of elements in the grammar, typically in dozens. So 32 is a reasonable number for not very complex schemas. If the grammar is extremely complex, more effective threads would be allowed, but for general purposes I think we can fix it to 32.

mgoin commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -87,11 +88,22 @@ def __init__(self, vllm_config: VllmConfig):
+        Returns:
+            int: The number of threads to use for the grammar compiler.
+        """
+        num_cpus = os.cpu_count()
+        return min(num_cpus / 2, 32)
```
Why are 32 threads needed for this? It seems like overkill and poses problems when running multiple instances on one host. 

This also won't work well for systems with small numbers of cores, as we have several threads being used just for multi-process input, scheduler, model workers. I think we need to have a fixed floor where you subtract some number of "reserved vllm threads" to get an "available threads" estimate

russellb replied:
if it's an 8 CPU core system, this will result in 4.

fair point about multiple instances on a single host.  See my comment on the PR with a test script that helped find the sweet spot . I don't think we should hard code 32 since there's nothing magic about that.

russellb replied:
@Ubospica why did you move away from using `psutil` and getting the number of physical cores? did you get an error?

mgoin replied:
basically I think we should increase the number of threads for structured output if we have an embarrassing number of cores, rather than taking up half by default

Ubospica replied:
> if it's an 8 CPU core system, this will result in 4.
> 
> fair point about multiple instances on a single host. See my comment on the PR with a test script that helped find the sweet spot . I don't think we should hard code 32 since there's nothing magic about that.

32 threads reflect the effective threads needed for typical schema sizes. This might be appropriate since we’re only setting an upper limit.

To allow multiple instances, I think we can add a global setting to limit the number of threads, just like we did in V1. But just not sure if we need to do it in this PR.

tmehlinger commented:
It would be better to make this configurable (say, `VLLM_XGRAMMER_COMPILER_THREADS`) rather than simply assuming a thread count based on number of available cores. If a config value isn't provided, _then_ fall back to your assumption. As it is, this has the potential to monopolize a substantial portion of available CPU time on a system that may be busy with other workloads, and operators have no way to tune the behavior.

russellb commented:
> It would be better to make this configurable (say, `VLLM_XGRAMMAR_COMPILER_THREADS`) rather than simply assuming a thread count based on number of available cores. If a config value isn't provided, _then_ fall back to your assumption. As it is, this has the potential to monopolize a substantial portion of available CPU time on a system that may be busy with other workloads, and operators have no way to tune the behavior.

This makes sense. The simplest non-controversial thing we could do right now is add the env var, set it to 8 by default, and document that it's a tunable worth exploring in some extreme cases.

Ubospica commented:
> > It would be better to make this configurable (say, `VLLM_XGRAMMAR_COMPILER_THREADS`) rather than simply assuming a thread count based on number of available cores. If a config value isn't provided, _then_ fall back to your assumption. As it is, this has the potential to monopolize a substantial portion of available CPU time on a system that may be busy with other workloads, and operators have no way to tune the behavior.
> 
> This makes sense. The simplest non-controversial thing we could do right now is add the env var, set it to 8 by default, and document that it's a tunable worth exploring in some extreme cases.

I agree we can have an env var. What about: if an env var is set, use it; otherwise, fallback to the method in this PR.

Ubospica committed:
use envvar

Signed-off-by: Ubospica <ubospica@gmail.com>

russellb commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -94,15 +96,26 @@ def __init__(self, vllm_config: VllmConfig):
             int: The number of threads to use for the grammar compiler.
         """
+        # check environment variable
+        env_threads = os.environ.get(VLLM_XGRAMMAR_COMPILER_THREADS_ENV_VAR)
```
We have another pattern we typically use for env vars in vLLM. I'm happy to do the conversion for you if you don't mind me committing to your branch. See `vllm.envs`

russellb commented:
> I agree we can have an env var. What about: if an env var is set, use it; otherwise, fallback to the method in this PR. This can achieve better performance while allowing control in extreme cases.

I think there is concern that we're increasing the number of threads when it may only provide a significant benefit in the more extreme cases. Instead, I suggest just keeping the existing default of 8, but allow tuning with the env var. I think that's the compromise that we can get merged.

russellb committed:
Move env var to `vllm.envs`

This changes the XGRAMMAR threads env var to be handled like other vllm
env vars.

This also removes the logic for auto calculating the thread pool size,
since that was controversial in review.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb reviewed (commented):
I've changed this to:

1. Change the env var handling to use `vllm.envs` like other vllm env vars

2. Remove the automatic calculation of the thread-pool size, since that part was controversial. It still defaults to `8`, but can now at least be changed using the env var if needed.

russellb committed:
Add notes on the xgrammar threads env var to the docs

Signed-off-by: Russell Bryant <rbryant@redhat.com>

Ubospica reviewed (commented):
LGTM!

Ubospica reviewed (commented):
@mgoin @russellb Is there any update with this PR?
